src/ty_check.rs

```diff
@@ -71,11 +72,22 @@ impl<'a> Context<'a> {
                         Box::new(Ty::Int),
                         Box::new(Ty::Arrow(Box::new(Ty::Int), Box::new(Ty::Bool))),
                     ),
-                    BinOp::Equal | BinOp::NotEqual => todo!(),
+                    BinOp::Equal | BinOp::NotEqual => Ty::Arrow(
+                        Box::new(Ty::Bool),
+                        Box::new(Ty::Arrow(Box::new(Ty::Bool), Box::new(Ty::Bool))),
+                    ),
                 },
             },
             Term::App(t1, t2) => {
-                let ty1 = self.type_of(t1)?;
+                let ty1 = if let box Term::App(box Term::Abs(Abstraction::Binary(_)), t3) = t1 {
```

Comment:
Here you need to check that the `op` inside  `Abstraction::Binary(op)` is `NotEqual` or `Equal`. Otherwise all the binary operations are being typed as `T -> T -> Bool`